### DISCUSSION TITLE
Proposal: Refactoring Spotbit and setting up CI pipelines

### DISCUSSION THREAD

**Discussion by HarshCasper** [BlockchainCommons/Community#53] [Ideas]
## Refactoring Spotbit

Spotbit is a Flask application designed to show-case Bitcoin price data and candles. My first milestone would revolve around publishing a documentation site for Spotbit to allow developers and general users, navigate across Spotbit and use it with maximum efficiency. 

The developer focussed documentation would revolve around: 

- Setup & Installation Instructions 
- API Documentation (Swagger UI)

The user-focussed documentation would revolve around: 

- Application workflow and design interface
- Managing configurations 
- Usage of exchanges for averaging

I would also be working on refactoring the `README.md` file to ensure that the documentation is modularized in effect. 

The milestones for this would be: 

- Setting up a Documentation Site using Sphinx 
- Setting up Swagger UI for the Flask API
- Refactoring and publishing documentation on the Sphinx
- Publishing the documentation over ReadTheDocs

| Milestone | Date |
| -------- | -------- |
| Setting up a Documentation Site using Sphinx     | July 10 |
| Setting up Swagger UI for the Flask API | July 13
| Refactoring and publishing documentation on the Sphinx | July 17
| Publishing the documentation over ReadTheDocs | July 20

Given the fact that I have prior experience in Sphinx, I can dually commit to the same.

Right now, the code quality of Spotbit's Flask API is at a lower end. I would be next working on refactoring the code to match the standards we would be looking for in code readability and maintenance. 

The milestones for this would be: 

- Setting up Code Quality checks using Flake8
- Refactoring the Codebase to suit the purpose of readability 
- Setup a CI pipeline for automated checks 

| Milestone | Date |
| -------- | -------- |
| Setting up Code Quality checks using Flake8     | July 22 |
| Refactoring the Codebase to suit the purpose of readability | July 26
| Setup a CI pipeline for automated checks | July 27

## Setting up CI pipelines 

I would also be working on developing consistent CI pipelines and project management tools and bots to be used with GitHub. 

This project would concern around the following: 

- [Consistent GitHub Repo labels](https://github.com/BlockchainCommons/Community/issues/39)
- [Local Electrum Server on Regtest](https://github.com/BlockchainCommons/sweeptool-cli/issues/2)
- [Adding GitHub Actions for Blockchain Commons Website](https://github.com/BlockchainCommons/www.blockchaincommons.com/issues/8)

I would also forward on identifying further areas where CI pipelines can be set up for automated checks. An extra area of focus would be on adding Issue/PR Templates across all of the repositories.

| Milestone | Date |
| -------- | -------- |
| Consistent GitHub Repo Labels     | July August 31 |
| Local Electrum Server on Regtest | August 3
| Adding GitHub Actions for Blockchain Commons Website | August 7

## Progress Made

I have successfully executed one of the Community Issues around [Sponsor Listing](https://github.com/BlockchainCommons/Community/issues/24). The [Pull Request](https://github.com/BlockchainCommons/www.blockchaincommons.com/pull/17) has been merged to achieve the same.

**Comment by gorazdko** [comment]
> Setting up Code Quality checks using Flake8

> Refactoring the Codebase to suit the purpose of readability

I'm not sure how useful that would be as you already pointed out correctly yourself:

> Right now, the code quality of Spotbit's Flask API is at a lower end.

I'd rather see spotbit rewritten from scratch instead of refactoring it. Not the whole project but only the essential part. Polling prices from a few different exchanges and making them available over spotbit via http request, i.e.  /now, /now/exchange (or sth basic like that)
Anything else above that (install script, testing over Tor etc. ) would be bonus features  or a new project for sb else

Here are some notes I made about the current status of spotbit:

## Abstract

There is a problem with spotbit, i.e. a server running spotbit gets clogged up (e.g. CPU rises to 100%)
which results in a server failure. This is caused by spotbit which ends up in a deadlocked state  when writing and reading from its database (this happens because spotbit is accessing db with multiple threads). This is not just a bug, but rather a flaw in architecture. So we need to redesign and rewrite spotbit from scratch.

Possible  solution:

Change architecture to: asyncio + flask like framework that supports asyncio

asyncio is essential to get rid of the threads
- [ ] first recreate what we have now with asyncio without using a DB. This will support
   commands such as /now /now/exchange
- [ ] secondly, introduce the DB (influxdb) if we want to additionally support somewhat historical prices

The first part is easier and essential whereas the second part is larger in scope (is considered a new project) and should be started with only after the first part is thoroughly tested and reviewed.
